internal/dataplane/parser/ingressrules.go

```diff
@@ -32,7 +34,8 @@ func mergeIngressRules(objs ...ingressRules) ingressRules {
 
 	for _, obj := range objs {
 		for k, v := range obj.SecretNameToSNIs {
-			result.SecretNameToSNIs[k] = append(result.SecretNameToSNIs[k], v...)
+			result.SecretNameToSNIs.addHosts(k, v.hosts)
```

Comment:
here we merges ALL `secretsNameToSNIs` into one, so calling `addHosts` for each ingress rule may call the makemap for thousands times and fall into the terrible performance. So I think #3166  would become a prerequisite of this because we merge ALL  `secretsNameToSNIs` for objects(`Ingress`,`HTTPRoute`,...) under control of KIC., which may have a large number.

Reply:
I switched back to the naive merging that we had before: 7087e89. Are you ok with merging this and tackling #3166 in a separate PR to not introduce too much noise here?

Reply:
I think it is OK because `mergeIngressRules` is not so slow and grows in a linear way.
```
func BenchmarkMergeIngressRules(b *testing.B) {
	for n := b.N; n > 0; n-- {
		inputs := []ingressRules{}
		ingressRuleNum := 100
		hostsPerIngress := 10
		for i := 0; i < ingressRuleNum; i++ {
			rule := newIngressRules()
			hosts := []string{}
			for j := 0; j < hostsPerIngress; j++ {
				hosts = append(hosts, "host-"+strconv.Itoa(i)+"-"+strconv.Itoa(j)+"example.com")
			}
			tlsSection := []netv1.IngressTLS{
				{Hosts: hosts, SecretName: "cert-" + strconv.Itoa(i)},
			}
			ing := &netv1.Ingress{
				ObjectMeta: metav1.ObjectMeta{
					Namespace: "default",
					Name:      "ing-" + strconv.Itoa(i),
				},
			}

			rule.SecretNameToSNIs.addFromIngressV1TLS(tlsSection, ing)
			inputs = append(inputs, rule)
		}
		mergeIngressRules(inputs...)
	}
}
```
It merges 100(and increased in following runs) ingress rules, each has 10 hosts sharing the same secret.
```
100: BenchmarkMergeIngressRules-8   	    4813	    251512 ns/op	  249770 B/op	    3318 allocs/op 
1000: BenchmarkMergeIngressRules-8   	     430	   2675415 ns/op	 2584781 B/op	   43854 allocs/op
10000: BenchmarkMergeIngressRules-8   	      43	  25789422 ns/op	25959571 B/op	  449100 allocs/op
```
If create a map each time, it grows `O(n^2)`.